File: src/commands/findbook.js
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const openaiService = require('../services/openaiService');
const bookService = require('../services/bookService');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('findbook')
    .setDescription('Find books based on description, title, author, or theme')
    .addStringOption(option =>
      option.setName('query')
        .setDescription('Describe the book you\'re looking for')
        .setRequired(true)),

  async execute(interaction) {
    await interaction.deferReply();
    
    try {
      const query = interaction.options.getString('query');
      
      // Let the AI parse the natural language query
      const searchParams = await openaiService.parseBookQuery(query);
      console.log('Parsed search parameters:', searchParams);
      
      // Search for books
      const books = await bookService.searchBooks(searchParams);
      
      if (books.length === 0) {
        // If no books found, use AI to provide a helpful response
        const aiResponse = await openaiService.enhanceBookResults([], query);
        await interaction.editReply(aiResponse);
        return;
      }
      
      // Get AI to enhance the book results with context
      const aiResponse = await openaiService.enhanceBookResults(books, query);
      
      // Create embeds for the books
      const embeds = books.slice(0, 3).map(book => {
        const embed = new EmbedBuilder()
          .setTitle(book.title)
          .setAuthor({ name: book.authors?.join(', ') || 'Unknown Author' })
          .setDescription(book.description?.substring(0, 300) + (book.description?.length > 300 ? '...' : '') || 'No description available')
          .addFields(
            { name: 'Published', value: book.publishedDate || 'Unknown', inline: true },
            { name: 'Categories', value: book.categories?.join(', ') || 'Unknown', inline: true }
          )
          .setColor(0x0099FF);
          
        if (book.imageLinks?.thumbnail) {
          embed.setThumbnail(book.imageLinks.thumbnail);
        }
        
        if (book.previewLink) {
          embed.setURL(book.previewLink);
        }
        
        return embed;
      });
      
      // Send the AI's response first
      await interaction.editReply({ content: aiResponse });
      
      // Then send the book embeds
      if (embeds.length > 0) {
        await interaction.followUp({ embeds });
      }
    } catch (error) {
      console.error('Error executing findbook command:', error);
      await interaction.editReply('Sorry, I encountered an error while searching for books. Please try again later.');
    }
  },
}; 

File: src/cogs/findbook.py
import discord
from discord import app_commands
from discord.ext import commands
import logging
from src.services.openai_service import OpenAIService
from src.services.book_service import BookService
from src.services.rag_service import RAGService

logger = logging.getLogger('bookfinder.commands.findbook')

class FindBookCog(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        
    @app_commands.command(
        name="findbook",
        description="Find books based on description, title, author, or theme"
    )
    @app_commands.describe(
        query="Describe the book you're looking for"
    )
    async def findbook(self, interaction: discord.Interaction, query: str):
        """
        Slash command to find books based on user's description
        """
        await interaction.response.defer()
        
        try:
            # Let the AI parse the natural language query
            search_params = await OpenAIService.parse_book_query(query)
            logger.info(f"Parsed search parameters: {search_params}")
            
            # Check if AI returned an error (for impossible queries)
            if "error" in search_params:
                error_message = search_params["error"]
                
                # Log the interaction with RAG
                RAGService.log_interaction(
                    user_id=interaction.user.id,
                    query=query,
                    books_found=[],
                    command_type="findbook",
                    response_text=error_message
                )
                
                await interaction.followup.send(error_message)
                return
            
            # Search for books
            books = await BookService.search_books(search_params)
            
            if not books:
                # If no books found, use AI to provide a helpful response
                ai_response = await OpenAIService.enhance_book_results([], query)
                
                # Log the interaction with RAG
                RAGService.log_interaction(
                    user_id=interaction.user.id,
                    query=query,
                    books_found=[],
                    command_type="findbook",
                    response_text=ai_response
                )
                
                await interaction.followup.send(ai_response)
                return
            
            # Get AI to enhance the book results with context
            ai_response = await OpenAIService.enhance_book_results(books, query)
            
            # Log the interaction with RAG BEFORE creating embeds
            RAGService.log_interaction(
                user_id=interaction.user.id,
                query=query,
                books_found=books,
                command_type="findbook",
                response_text=ai_response
            )
            
            # Create embeds for the books
            embeds = []
            for book in books[:3]:  # Limit to top 3 books
                embed = discord.Embed(
                    title=book['title'],
                    description=book.get('description', 'No description available')[:300] + 
                               ('...' if book.get('description', '') and len(book['description']) > 300 else ''),
                    color=discord.Color.blue(),
                    url=book.get('previewLink', '')
                )
                
                embed.set_author(name=", ".join(book.get('authors', ['Unknown Author'])))
                
                # Add fields
                embed.add_field(
                    name="Published", 
                    value=book.get('publishedDate', 'Unknown'), 
                    inline=True
                )
                
                if book.get('categories'):
                    embed.add_field(
                        name="Categories", 
                        value=", ".join(book['categories']), 
                        inline=True
                    )
                
                # Add thumbnail if available
                if book.get('imageLinks', {}).get('thumbnail'):
                    embed.set_thumbnail(url=book['imageLinks']['thumbnail'])
                
                embeds.append(embed)
            
            # Send the AI's response first
            await interaction.followup.send(content=ai_response)
            
            # Then send the book embeds
            if embeds:
                await interaction.followup.send(embeds=embeds)
                
        except Exception as e:
            logger.error(f"Error executing findbook command: {e}")
            
            # Log the error interaction
            RAGService.log_interaction(
                user_id=interaction.user.id,
                query=query,
                books_found=[],
                command_type="findbook",
                response_text="Error occurred"
            )
            
            await interaction.followup.send(
                "Sorry, I encountered an error while searching for books. Please try again later."
            )

async def setup(bot):
    await bot.add_cog(FindBookCog(bot)) 